File: shared/net_schema.py
from __future__ import annotations

from typing import Any


SNAPSHOT_SCHEMA = "compact-v1"

_COLLECTION_TO_WIRE = {
    "players": "p",
    "zombies": "z",
    "projectiles": "s",
    "grenades": "g",
    "mines": "m",
    "poison_projectiles": "pp",
    "poison_pools": "pl",
    "loot": "l",
    "buildings": "b",
}
_WIRE_TO_COLLECTION = {wire: collection for collection, wire in _COLLECTION_TO_WIRE.items()}


def compact_snapshot(snapshot: dict[str, Any], local_player_id: str | None) -> dict[str, Any]:
    players = _collection(snapshot, "players")
    local_player = players.get(local_player_id or "")
    compact: dict[str, Any] = {
        "v": 1,
        "t": snapshot.get("time", 0.0),
        "mw": snapshot.get("map_width", 1),
        "mh": snapshot.get("map_height", 1),
        "p": [_pack_player(entity) for entity_id, entity in players.items() if entity_id != local_player_id],
        "z": [_pack_zombie(entity) for entity in _collection(snapshot, "zombies").values()],
        "s": [_pack_projectile(entity) for entity in _collection(snapshot, "projectiles").values()],
        "g": [_pack_grenade(entity) for entity in _collection(snapshot, "grenades").values()],
        "m": [_pack_mine(entity) for entity in _collection(snapshot, "mines").values()],
        "pp": [_pack_poison_projectile(entity) for entity in _collection(snapshot, "poison_projectiles").values()],
        "pl": [_pack_poison_pool(entity) for entity in _collection(snapshot, "poison_pools").values()],
        "l": [_pack_loot(entity) for entity in _collection(snapshot, "loot").values()],
        "b": _collection(snapshot, "buildings"),
    }
    if isinstance(local_player, dict):
        compact["lp"] = local_player
    return compact


def expand_snapshot(payload: dict[str, Any]) -> dict[str, Any]:
    if payload.get("v") != 1:
        return payload
    players = {}
    local_player = payload.get("lp")
    if isinstance(local_player, dict) and "id" in local_player:
        players[str(local_player["id"])] = local_player
    for row in payload.get("p", []):
        player = _unpack_player(row)
        players[player["id"]] = player
    return {
        "time": float(payload.get("t", 0.0)),
        "map_width": int(payload.get("mw", 1)),
        "map_height": int(payload.get("mh", 1)),
        "players": players,
        "zombies": _unpack_rows(payload.get("z", []), _unpack_zombie),
        "projectiles": _unpack_rows(payload.get("s", []), _unpack_projectile),
        "grenades": _unpack_rows(payload.get("g", []), _unpack_grenade),
        "mines": _unpack_rows(payload.get("m", []), _unpack_mine),
        "poison_projectiles": _unpack_rows(payload.get("pp", []), _unpack_poison_projectile),
        "poison_pools": _unpack_rows(payload.get("pl", []), _unpack_poison_pool),
        "loot": _unpack_rows(payload.get("l", []), _unpack_loot),
        "buildings": _collection(payload, "b"),
    }


def compact_delta(delta: dict[str, Any], local_player_id: str | None) -> dict[str, Any]:
    compact: dict[str, Any] = {"v": 1}
    if "time" in delta:
        compact["t"] = delta["time"]
    if "map_width" in delta:
        compact["mw"] = delta["map_width"]
    if "map_height" in delta:
        compact["mh"] = delta["map_height"]
    for collection, wire_key in _COLLECTION_TO_WIRE.items():
        patch = delta.get(collection)
        if not isinstance(patch, dict):
            continue
        upserts = _collection(patch, "u")
        removals = list(patch.get("r", []))
        if collection == "players":
            full_players = {
                entity_id: entity
                for entity_id, entity in upserts.items()
                if entity_id == local_player_id or _is_full_player(entity)
            }
            rows = [_pack_player(entity) for entity_id, entity in upserts.items() if entity_id not in full_players]
            compact_patch: dict[str, Any] = {"u": rows, "r": removals}
            if full_players:
                compact_patch["f"] = full_players
        elif collection == "buildings":
            compact_patch = {"u": upserts, "r": removals}
        else:
            compact_patch = {"u": [_pack_entity(collection, entity) for entity in upserts.values()], "r": removals}
        if compact_patch.get("u") or compact_patch.get("f") or compact_patch.get("r"):
            compact[wire_key] = compact_patch
    return compact


def expand_delta(payload: dict[str, Any]) -> dict[str, Any]:
    if payload.get("v") != 1:
        return payload
    delta: dict[str, Any] = {}
    if "t" in payload:
        delta["time"] = payload["t"]
    if "mw" in payload:
        delta["map_width"] = payload["mw"]
    if "mh" in payload:
        delta["map_height"] = payload["mh"]
    for wire_key, collection in _WIRE_TO_COLLECTION.items():
        patch = payload.get(wire_key)
        if not isinstance(patch, dict):
            continue
        upserts: dict[str, Any] = {}
        if collection == "players":
            for row in patch.get("u", []):
                entity = _unpack_player(row)
                upserts[entity["id"]] = entity
            for entity_id, entity in _collection(patch, "f").items():
                upserts[entity_id] = entity
        elif collection == "buildings":
            upserts.update(_collection(patch, "u"))
        else:
            for row in patch.get("u", []):
                entity = _unpack_entity(collection, row)
                upserts[entity["id"]] = entity
        delta[collection] = {"u": upserts, "r": list(patch.get("r", []))}
    return delta


def _pack_entity(collection: str, entity: dict[str, Any]) -> list[Any]:
    if collection == "zombies":
        return _pack_zombie(entity)
    if collection == "projectiles":
        return _pack_projectile(entity)
    if collection == "grenades":
        return _pack_grenade(entity)
    if collection == "mines":
        return _pack_mine(entity)
    if collection == "poison_projectiles":
        return _pack_poison_projectile(entity)
    if collection == "poison_pools":
        return _pack_poison_pool(entity)
    if collection == "loot":
        return _pack_loot(entity)
    return [entity.get("id", "")]


def _unpack_entity(collection: str, row: list[Any]) -> dict[str, Any]:
    if collection == "zombies":
        return _unpack_zombie(row)
    if collection == "projectiles":
        return _unpack_projectile(row)
    if collection == "grenades":
        return _unpack_grenade(row)
    if collection == "mines":
        return _unpack_mine(row)
    if collection == "poison_projectiles":
        return _unpack_poison_projectile(row)
    if collection == "poison_pools":
        return _unpack_poison_pool(row)
    if collection == "loot":
        return _unpack_loot(row)
    return {"id": str(row[0]) if row else ""}


def _pack_player(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    return [
        entity.get("id", ""),
        entity.get("name", "Player"),
        _q(pos.get("x")),
        _q(pos.get("y")),
        _q(entity.get("angle"), 1000),
        int(entity.get("health", 100)),
        int(entity.get("armor", 0)),
        entity.get("armor_key", "none"),
        entity.get("active_slot", "1"),
        1 if entity.get("alive", True) else 0,
        int(entity.get("score", 0)),
        entity.get("kills_by_kind", {}),
        _q(entity.get("noise"), 10),
        1 if entity.get("sprinting", False) else 0,
        int(entity.get("floor", 0)),
        entity.get("inside_building"),
        1 if entity.get("sneaking", False) else 0,
        _q(entity.get("poison_left"), 1000),
        _active_weapon_payload(entity),
        int(entity.get("ping_ms", 0)),
        entity.get("connection_quality", "stable"),
    ]


def _unpack_player(row: list[Any]) -> dict[str, Any]:
    active_slot = str(_get(row, 8, "1"))
    weapon = _get(row, 18, None)
    weapons = {active_slot: weapon} if isinstance(weapon, dict) else {}
    return {
        "id": str(_get(row, 0, "")),
        "name": str(_get(row, 1, "Player")),
        "pos": {"x": _uq(_get(row, 2, 0)), "y": _uq(_get(row, 3, 0))},
        "angle": _uq(_get(row, 4, 0), 1000),
        "health": int(_get(row, 5, 100)),
        "armor": int(_get(row, 6, 0)),
        "armor_key": str(_get(row, 7, "none")),
        "active_slot": active_slot,
        "alive": bool(_get(row, 9, 1)),
        "score": int(_get(row, 10, 0)),
        "kills_by_kind": _get(row, 11, {}),
        "noise": _uq(_get(row, 12, 0), 10),
        "sprinting": bool(_get(row, 13, 0)),
        "floor": int(_get(row, 14, 0)),
        "inside_building": _get(row, 15, None),
        "sneaking": bool(_get(row, 16, 0)),
        "poison_left": _uq(_get(row, 17, 0), 1000),
        "weapons": weapons,
        "ping_ms": int(_get(row, 19, 0)),
        "connection_quality": str(_get(row, 20, "stable")),
    }


def _pack_zombie(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    return [
        entity.get("id", ""),
        entity.get("kind", "walker"),
        _q(pos.get("x")),
        _q(pos.get("y")),
        int(entity.get("health", 1)),
        int(entity.get("armor", 0)),
        entity.get("mode", "patrol"),
        _q(entity.get("facing"), 1000),
        int(entity.get("floor", 0)),
        entity.get("target_player_id"),
        _q(entity.get("alertness"), 1000),
    ]


def _unpack_zombie(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "kind": str(_get(row, 1, "walker")),
        "pos": {"x": _uq(_get(row, 2, 0)), "y": _uq(_get(row, 3, 0))},
        "health": int(_get(row, 4, 1)),
        "armor": int(_get(row, 5, 0)),
        "mode": str(_get(row, 6, "patrol")),
        "facing": _uq(_get(row, 7, 0), 1000),
        "floor": int(_get(row, 8, 0)),
        "target_player_id": _get(row, 9, None),
        "alertness": _uq(_get(row, 10, 0), 1000),
    }


def _pack_projectile(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    vel = _velocity(entity)
    return [
        entity.get("id", ""),
        entity.get("owner_id", ""),
        _q(pos.get("x")),
        _q(pos.get("y")),
        _q(vel.get("x")),
        _q(vel.get("y")),
        int(entity.get("damage", 1)),
        _q(entity.get("life"), 1000),
        _q(entity.get("radius", 5.0)),
        int(entity.get("floor", 0)),
    ]


def _unpack_projectile(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "owner_id": str(_get(row, 1, "")),
        "pos": {"x": _uq(_get(row, 2, 0)), "y": _uq(_get(row, 3, 0))},
        "velocity": {"x": _uq(_get(row, 4, 0)), "y": _uq(_get(row, 5, 0))},
        "damage": int(_get(row, 6, 1)),
        "life": _uq(_get(row, 7, 0), 1000),
        "radius": _uq(_get(row, 8, 50)),
        "floor": int(_get(row, 9, 0)),
    }


def _pack_grenade(entity: dict[str, Any]) -> list[Any]:
    row = _pack_projectile({**entity, "damage": 0, "life": entity.get("timer", 0.0)})
    row.extend([entity.get("kind", "grenade")])
    return row


def _unpack_grenade(row: list[Any]) -> dict[str, Any]:
    projectile = _unpack_projectile(row)
    return {
        "id": projectile["id"],
        "owner_id": projectile["owner_id"],
        "pos": projectile["pos"],
        "velocity": projectile["velocity"],
        "timer": projectile["life"],
        "floor": projectile["floor"],
        "radius": projectile["radius"],
        "kind": str(_get(row, 10, "grenade")),
    }


def _pack_mine(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    return [
        entity.get("id", ""),
        entity.get("owner_id", ""),
        entity.get("kind", "mine_standard"),
        _q(pos.get("x")),
        _q(pos.get("y")),
        int(entity.get("floor", 0)),
        1 if entity.get("armed", False) else 0,
        _q(entity.get("trigger_radius", 100.0)),
        _q(entity.get("blast_radius", 220.0)),
        _q(entity.get("rotation", 0.0), 1000),
    ]


def _unpack_mine(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "owner_id": str(_get(row, 1, "")),
        "kind": str(_get(row, 2, "mine_standard")),
        "pos": {"x": _uq(_get(row, 3, 0)), "y": _uq(_get(row, 4, 0))},
        "floor": int(_get(row, 5, 0)),
        "armed": bool(_get(row, 6, 0)),
        "trigger_radius": _uq(_get(row, 7, 1000)),
        "blast_radius": _uq(_get(row, 8, 2200)),
        "rotation": _uq(_get(row, 9, 0), 1000),
    }


def _pack_poison_projectile(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    vel = _velocity(entity)
    target = entity.get("target", {}) if isinstance(entity.get("target"), dict) else {}
    return [
        entity.get("id", ""),
        entity.get("owner_id", ""),
        _q(pos.get("x")),
        _q(pos.get("y")),
        _q(vel.get("x")),
        _q(vel.get("y")),
        _q(target.get("x")),
        _q(target.get("y")),
        int(entity.get("floor", 0)),
        _q(entity.get("radius", 9.0)),
        _q(entity.get("life", 2.4), 1000),
    ]


def _unpack_poison_projectile(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "owner_id": str(_get(row, 1, "")),
        "pos": {"x": _uq(_get(row, 2, 0)), "y": _uq(_get(row, 3, 0))},
        "velocity": {"x": _uq(_get(row, 4, 0)), "y": _uq(_get(row, 5, 0))},
        "target": {"x": _uq(_get(row, 6, 0)), "y": _uq(_get(row, 7, 0))},
        "floor": int(_get(row, 8, 0)),
        "radius": _uq(_get(row, 9, 90)),
        "life": _uq(_get(row, 10, 2400), 1000),
    }


def _pack_poison_pool(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    return [
        entity.get("id", ""),
        _q(pos.get("x")),
        _q(pos.get("y")),
        int(entity.get("floor", 0)),
        _q(entity.get("timer", 5.0), 1000),
        _q(entity.get("radius", 54.0)),
    ]


def _unpack_poison_pool(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "pos": {"x": _uq(_get(row, 1, 0)), "y": _uq(_get(row, 2, 0))},
        "floor": int(_get(row, 3, 0)),
        "timer": _uq(_get(row, 4, 5000), 1000),
        "radius": _uq(_get(row, 5, 540)),
    }


def _pack_loot(entity: dict[str, Any]) -> list[Any]:
    pos = _pos(entity)
    return [
        entity.get("id", ""),
        entity.get("kind", "item"),
        _q(pos.get("x")),
        _q(pos.get("y")),
        entity.get("payload", ""),
        int(entity.get("amount", 1)),
        int(entity.get("floor", 0)),
        entity.get("rarity", "common"),
    ]


def _unpack_loot(row: list[Any]) -> dict[str, Any]:
    return {
        "id": str(_get(row, 0, "")),
        "kind": str(_get(row, 1, "item")),
        "pos": {"x": _uq(_get(row, 2, 0)), "y": _uq(_get(row, 3, 0))},
        "payload": str(_get(row, 4, "")),
        "amount": int(_get(row, 5, 1)),
        "floor": int(_get(row, 6, 0)),
        "rarity": str(_get(row, 7, "common")),
    }


def _unpack_rows(rows: Any, unpacker: Any) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for row in rows if isinstance(rows, list) else []:
        entity = unpacker(row)
        result[entity["id"]] = entity
    return result


def _active_weapon_payload(entity: dict[str, Any]) -> dict[str, Any] | None:
    active_slot = str(entity.get("active_slot", "1"))
    weapons = entity.get("weapons")
    if not isinstance(weapons, dict):
        return None
    weapon = weapons.get(active_slot)
    return weapon if isinstance(weapon, dict) else None


def _is_full_player(entity: dict[str, Any]) -> bool:
    return any(key in entity for key in ("backpack", "equipment", "quick_items", "owned_armors", "medkits"))


def _collection(data: dict[str, Any], key: str) -> dict[str, Any]:
    value = data.get(key, {})
    return value if isinstance(value, dict) else {}


def _pos(entity: dict[str, Any]) -> dict[str, Any]:
    pos = entity.get("pos", {})
    return pos if isinstance(pos, dict) else {}


def _velocity(entity: dict[str, Any]) -> dict[str, Any]:
    velocity = entity.get("velocity", {})
    return velocity if isinstance(velocity, dict) else {}


def _q(value: Any, scale: int = 10) -> int:
    try:
        return int(round(float(value) * scale))
    except (TypeError, ValueError):
        return 0


def _uq(value: Any, scale: int = 10) -> float:
    try:
        return round(float(value) / scale, 3)
    except (TypeError, ValueError):
        return 0.0


def _get(row: list[Any], index: int, fallback: Any) -> Any:
    return row[index] if isinstance(row, list) and index < len(row) else fallback

File: server_go/shared/net_schema.go
package shared

import "math"

const SnapshotSchema = "compact-v1"

func CompactSnapshot(snapshot WorldSnapshot, localPlayerID string, interestRadius float64) map[string]any {
	local := snapshot.Players[localPlayerID]
	compact := map[string]any{
		"v":  1,
		"t":  Round(snapshot.Time, 3),
		"mw": int(snapshot.MapWidth),
		"mh": int(snapshot.MapHeight),
		"p":  []any{},
		"z":  []any{},
		"s":  []any{},
		"g":  []any{},
		"m":  []any{},
		"pp": []any{},
		"pl": []any{},
		"l":  []any{},
		"b":  buildingsPayload(snapshot.Buildings),
	}
	if local != nil {
		compact["lp"] = PlayerDict(local)
	}
	compact["p"] = filterPackPlayers(snapshot.Players, local, localPlayerID, interestRadius)
	compact["z"] = filterPackZombies(snapshot.Zombies, local, interestRadius)
	compact["s"] = filterPackProjectiles(snapshot.Projectiles, local, interestRadius)
	compact["g"] = filterPackGrenades(snapshot.Grenades, local, interestRadius)
	compact["m"] = filterPackMines(snapshot.Mines, local, interestRadius)
	compact["pp"] = filterPackPoisonProjectiles(snapshot.PoisonProjectiles, local, interestRadius)
	compact["pl"] = filterPackPoisonPools(snapshot.PoisonPools, local, interestRadius)
	compact["l"] = filterPackLoot(snapshot.Loot, local, interestRadius)
	return compact
}

func PlayerDict(p *PlayerState) map[string]any {
	weapons := map[string]any{}
	for slot, weapon := range p.Weapons {
		weapons[slot] = WeaponDict(weapon)
	}
	backpack := make([]any, len(p.Backpack))
	for i, item := range p.Backpack {
		if item != nil {
			backpack[i] = item.ToMap()
		}
	}
	equipment := map[string]any{"head": nil, "torso": nil, "legs": nil, "arms": nil}
	for slot, item := range p.Equipment {
		if item != nil {
			equipment[slot] = item.ToMap()
		} else {
			equipment[slot] = nil
		}
	}
	quickItems := map[string]any{}
	for slot, item := range p.QuickItems {
		if item != nil {
			quickItems[slot] = item.ToMap()
		} else {
			quickItems[slot] = nil
		}
	}
	return map[string]any{
		"id":                 p.ID,
		"name":               p.Name,
		"pos":                p.Pos.ToMap(),
		"angle":              Round(p.Angle, 4),
		"health":             p.Health,
		"armor":              p.Armor,
		"armor_key":          p.ArmorKey,
		"speed":              p.Speed,
		"active_slot":        p.ActiveSlot,
		"alive":              p.Alive,
		"score":              p.Score,
		"kills_by_kind":      p.KillsByKind,
		"medkits":            p.Medkits,
		"owned_armors":       p.OwnedArmors,
		"noise":              Round(p.Noise, 3),
		"sprinting":          p.Sprinting,
		"floor":              p.Floor,
		"inside_building":    p.InsideBuilding,
		"sneaking":           p.Sneaking,
		"backpack":           backpack,
		"equipment":          equipment,
		"quick_items":        quickItems,
		"healing_left":       Round(p.HealingLeft, 3),
		"healing_rate":       Round(p.HealingRate, 3),
		"healing_pool":       Round(p.HealingPool, 3),
		"healing_stacks":     p.HealingStacks,
		"poison_left":        Round(p.PoisonLeft, 3),
		"poison_tick":        Round(p.PoisonTick, 3),
		"poison_damage":      p.PoisonDamage,
		"melee_cooldown":     Round(p.MeleeCooldown, 3),
		"notice":             p.Notice,
		"notice_timer":       Round(p.NoticeTimer, 3),
		"ping_ms":            p.PingMS,
		"connection_quality": p.ConnectionQuality,
		"weapons":            weapons,
	}
}

func WeaponDict(w *WeaponRuntime) map[string]any {
	if w == nil {
		return nil
	}
	return w.ToMap()
}

func PackPlayer(p *PlayerState) []any {
	return []any{
		p.ID,
		p.Name,
		q(p.Pos.X, 10),
		q(p.Pos.Y, 10),
		q(p.Angle, 1000),
		p.Health,
		p.Armor,
		p.ArmorKey,
		p.ActiveSlot,
		boolInt(p.Alive),
		p.Score,
		p.KillsByKind,
		q(p.Noise, 10),
		boolInt(p.Sprinting),
		p.Floor,
		nilIfEmptyAny(p.InsideBuilding),
		boolInt(p.Sneaking),
		q(p.PoisonLeft, 1000),
		WeaponDict(p.ActiveWeapon()),
		p.PingMS,
		p.ConnectionQuality,
	}
}

func PackZombie(z *ZombieState) []any {
	return []any{
		z.ID,
		z.Kind,
		q(z.Pos.X, 10),
		q(z.Pos.Y, 10),
		z.Health,
		z.Armor,
		z.Mode,
		q(z.Facing, 1000),
		z.Floor,
		nilIfEmpty(z.TargetPlayerID),
		q(z.Alertness, 1000),
	}
}

func PackProjectile(p *ProjectileState) []any {
	return []any{
		p.ID,
		p.OwnerID,
		q(p.Pos.X, 10),
		q(p.Pos.Y, 10),
		q(p.Velocity.X, 10),
		q(p.Velocity.Y, 10),
		p.Damage,
		q(p.Life, 1000),
		q(p.Radius, 10),
		p.Floor,
		p.WeaponKey,
	}
}

func PackGrenade(g *GrenadeState) []any {
	return []any{
		g.ID,
		g.OwnerID,
		q(g.Pos.X, 10),
		q(g.Pos.Y, 10),
		q(g.Velocity.X, 10),
		q(g.Velocity.Y, 10),
		0,
		q(g.Timer, 1000),
		q(g.Radius, 10),
		g.Floor,
		g.Kind,
	}
}

func PackMine(m *MineState) []any {
	return []any{
		m.ID,
		m.OwnerID,
		m.Kind,
		q(m.Pos.X, 10),
		q(m.Pos.Y, 10),
		m.Floor,
		boolInt(m.Armed),
		q(m.TriggerRadius, 10),
		q(m.BlastRadius, 10),
		q(m.Rotation, 1000),
	}
}

func PackPoisonProjectile(p *PoisonProjectileState) []any {
	return []any{
		p.ID,
		p.OwnerID,
		q(p.Pos.X, 10),
		q(p.Pos.Y, 10),
		q(p.Velocity.X, 10),
		q(p.Velocity.Y, 10),
		q(p.Target.X, 10),
		q(p.Target.Y, 10),
		p.Floor,
		q(p.Radius, 10),
		q(p.Life, 1000),
	}
}

func PackPoisonPool(p *PoisonPoolState) []any {
	return []any{
		p.ID,
		q(p.Pos.X, 10),
		q(p.Pos.Y, 10),
		p.Floor,
		q(p.Timer, 1000),
		q(p.Radius, 10),
	}
}

func PackLoot(l *LootState) []any {
	return []any{
		l.ID,
		l.Kind,
		q(l.Pos.X, 10),
		q(l.Pos.Y, 10),
		l.Payload,
		l.Amount,
		l.Floor,
		l.Rarity,
	}
}

func filterPackPlayers(players map[string]*PlayerState, local *PlayerState, localID string, radius float64) []any {
	rows := make([]any, 0, len(players))
	for id, player := range players {
		if id == localID {
			continue
		}
		if !visibleToLocal(local, player.Pos, player.Floor, radius) {
			continue
		}
		rows = append(rows, PackPlayer(player))
	}
	return rows
}

func filterPackZombies(zombies map[string]*ZombieState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(zombies))
	for _, zombie := range zombies {
		if !visibleToLocal(local, zombie.Pos, zombie.Floor, radius) {
			continue
		}
		rows = append(rows, PackZombie(zombie))
	}
	return rows
}

func filterPackProjectiles(projectiles map[string]*ProjectileState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(projectiles))
	for _, projectile := range projectiles {
		if !visibleToLocal(local, projectile.Pos, projectile.Floor, radius) {
			continue
		}
		rows = append(rows, PackProjectile(projectile))
	}
	return rows
}

func filterPackGrenades(grenades map[string]*GrenadeState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(grenades))
	for _, grenade := range grenades {
		if !visibleToLocal(local, grenade.Pos, grenade.Floor, radius) {
			continue
		}
		rows = append(rows, PackGrenade(grenade))
	}
	return rows
}

func filterPackMines(mines map[string]*MineState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(mines))
	for _, mine := range mines {
		if !visibleToLocal(local, mine.Pos, mine.Floor, radius) {
			continue
		}
		rows = append(rows, PackMine(mine))
	}
	return rows
}

func filterPackPoisonProjectiles(projectiles map[string]*PoisonProjectileState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(projectiles))
	for _, projectile := range projectiles {
		if !visibleToLocal(local, projectile.Pos, projectile.Floor, radius) {
			continue
		}
		rows = append(rows, PackPoisonProjectile(projectile))
	}
	return rows
}

func filterPackPoisonPools(pools map[string]*PoisonPoolState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(pools))
	for _, pool := range pools {
		if !visibleToLocal(local, pool.Pos, pool.Floor, radius) {
			continue
		}
		rows = append(rows, PackPoisonPool(pool))
	}
	return rows
}

func filterPackLoot(loot map[string]*LootState, local *PlayerState, radius float64) []any {
	rows := make([]any, 0, len(loot))
	for _, item := range loot {
		if !visibleToLocal(local, item.Pos, item.Floor, radius) {
			continue
		}
		rows = append(rows, PackLoot(item))
	}
	return rows
}

func visibleToLocal(local *PlayerState, pos Vec2, floor int, radius float64) bool {
	if local == nil {
		return true
	}
	if local.Floor != floor {
		return false
	}
	return local.Pos.Distance(pos) <= radius
}

func buildingsPayload(buildings map[string]*BuildingState) map[string]any {
	out := make(map[string]any, len(buildings))
	for id, building := range buildings {
		out[id] = building.ToMap()
	}
	return out
}

func q(value float64, scale float64) int {
	return int(math.Round(value * scale))
}

func boolInt(value bool) int {
	if value {
		return 1
	}
	return 0
}

func nilIfEmpty(value string) any {
	if value == "" {
		return nil
	}
	return value
}

func nilIfEmptyAny(value any) any {
	if value == nil {
		return nil
	}
	if text, ok := value.(string); ok && text == "" {
		return nil
	}
	return value
}
